Openbrush Grant Application from Supercolony

themarkian opened the pull request:
# Grant Application Checklist

- [x] The [application template](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI) address for the payment of the milestones is provided inside the application.  
- [x] I have read and acknowledged the [Terms and Conditions](https://github.com/w3f/Open-Grants-Program/blob/master/src/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The total funding amount of the project is below USD $30k for initial grant applications and $100k for follow-up grants.
- [x] The initial PR contains only one commit (squash if needed before submitting your PR).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/w3f/Open-Grants-Program?pullRequest=497) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/w3f/Open-Grants-Program?pullRequest=497) before we can accept your contribution.<br/>**0** out of **4** committers have signed the CLA.<br/><br/>:x: KseniaSupercolony<br/>:x: themarkian<br/>:x: xgreenx<br/>:x: Agamimosh<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/w3f/Open-Grants-Program?pullRequest=497) it.</sub>

alxs commented:
Thanks for the application @xgreenx, looks great. I think it's safe to say we're highly interested in this. If I remember correctly you also applied or at least were planning to apply for treasury funding, could you link to to the proposal for reference? What was the outcome?

themarkian commented:
Hey @alxs!
Thanks a lot for such quick feedback. Appreciate that a lot!
Yes, we considered internally applying for a treasury, but never did it.

alxs commented:
I see, I was confusing this with Patract Labs' [Metis proposal](https://polkadot.polkassembly.io/post/469). Are you aware of that project and have you been in contact with them? It'd in any case be useful to have alternative implementations, but maybe you can share resource or collaborate with them in some aspects.

Other than that I don't have a lot to add. At least from my side I'd be happy to go ahead with the application. Could you just make the milestones/costs clearer with regards to the current proposal? I'd recommend you split the milestones into previous work, scope of this grant and future plans. Make sure that the milestone(s) for this grant add up to $30k and feel free to omit the costs elsewhere.

themarkian commented:
@alxs yes, we are aware and formed a partnership with them recently. We intend to collaborate together towards our common goal which is in building an ecosystem. However, with metis they see things a bit differently than we see with openbrush. For example, openbrush isn't restricted to only providing the base implementation of contracts.  On top of that our project is in a more advanced stage currently and ready to be released as soon as the PSP-20(ERC20 for Polkadot) standard has reached a mature state. Supercolony currently leads the process and your help with it would be much appreciated 

More info about how we are different in metis is in the application form here - https://github.com/Supercolony-net/Open-Grants-Program/blob/master/applications/openbrush.md#ecosystem-fit

themarkian commented:
@alxs could you elaborate on what kind of changes are requested?

themarkian commented:
>could you just make the milestones/costs clearer with regards to the current proposal? I'd recommend you split the milestones into previous work, scope of this grant and future plans. Make sure that the milestone(s) for this grant add up to $30k and feel free to omit the costs elsewhere.

@alxs all right, would you also recommend that we request costs only for future work or grant could cover some of the work that was done already?

alxs commented:
I think this is a bit of a gray area at the moment. I'm personally in favour of only awarding grants for future work, but IIRC we've made exceptions in the past. It's probably best to check with @Noc2. Personally I'd be happy to accept a grant for, say, 50% past and 50% future work in this particular case, assuming we have no hard rule on the matter.

themarkian commented:
Hey @alxs!
Change has been implemented. We divided milestones into 3 sections - previous work, the scope of this grant, and future work. And the cost for the scope of this grant is estimated at 30k $.

alxs commented:
Thanks, looks good! If you want to deliver M1 with this grant also (in order to received feedback and so you can announce the grant as soon as it is approved) feel free to include it in the scope at no charge. Or split the $30k among M1/M2 as I said, both are fine with me.

Also in general feel free to work on ink! itself as part of a grant (I see you've included it in some of the deliverables, just wanted to make it clear that this is perfectly fine and even encouraged from our side).

themarkian commented:
Thanks, @alxs!
yeah, we were definitely planning to contribute to ink! and substrate. Ink! has already merged a couple of our PRs. M1 is included in the scope under the previous work section with no charge :) 

So what will be our next steps to get this grant approved?

alxs commented:
Of course, what I mean is that you may want to deliver it as a milestone anyway (i.e. move it into the scope of this grant, even though officially we won't compensate you for it) with the advantages that I mentioned before. This is completely up to you :)

I'm happy to share your application with the rest of the committee as soon as you decide on the above.

themarkian commented:
@alxs updated an application accordingly to include M1 in the scope. Looking forward to the feedback from the rest of the committee :)

alxs commented on applications/openbrush.md:
```diff
@@ -0,0 +1,274 @@
+
+* **Estimated duration:** 2 weeks
+* **FTE:**  2.5
+* **Costs:** 10,000 USD (this cost is covered by Supercolony)
```
```suggestion
* **Costs:** ~~10,000 USD~~ (this cost is covered by Supercolony)
```

alxs commented on applications/openbrush.md:
```diff
@@ -0,0 +1,273 @@
+| 0a.    | License                                             | MIT                                                                                                                                                                                              |
+| 0b.    | Documentation                                       | We will enhance inline documentation, update previous documentation based on simplifications, documentation for new features.                                                                                        |
+| 0c.    | Testing Guide                                       | We will test macros, update tests according to new features, simplifications.                                                                                               |
+| 1.     | Remove boilerplate code                             | We will provide a macro which will allow to remove boilerplate during usage of library(Library provides implementation on rust level in internal trait. User must reuse internal implementation with ink! messages. Our macro will simplify it). It will simplify the code structure and usage. |
```
Items 1-3 here were listed as achieved in the first version of your application. Could you clarify?

themarkian replied:
Indeed. 
However, they are currently under reconstruction for a few reasons, mainly coz we found problems with the current approach. so they need to be optimized.

themarkian replied:
I can go into details if you like

semuelle reviewed (dismissed):
I was not aware of this project until now. Looking forward to it!

Noc2 reviewed (changes requested):
In general the application looks really good and interesting. Regarding paying for previous work, this is something which we don’t do. Given that you were also able to work on it without a grant/our money. However, we have paid small amounts for teams in the past in order to open source (and improve) their existing work, which is obviously valuable for the entire ecosystem. Independent of it showing previous work, makes the grant itself most of the time more valuable for us/gives it a higher chance to be accepted. 
I also have a few questions:
- Is your milestone 1 already open source and if so could you share a link? 
- Are you aware of https://github.com/paritytech/substrate/issues/9354, which might lead to some changes regarding ink!? 
- How are your implementations of milestone 1 different from https://github.com/paritytech/ink/tree/master/examples?

themarkian commented:
Hey @Noc2!
thanks for reviewing our application. Glad to answer those questions. 

So currently only M1 is fully finished. M2 is what we are working on at the moment. 

1) Yes, it is open-sourced from the beginning. You can find the implementation of contracts here:
https://github.com/Supercolony-net/openbrush-contracts/tree/main/contracts
You can check the roadmap and releases for more details.
Roadmap: https://github.com/Supercolony-net/openbrush-contracts#roadmap-
Releases: https://github.com/Supercolony-net/openbrush-contracts/releases

2) Yes, we are aware of and interested in this issue because ink!'s contracts take really a lot of space. We are ready for any changes and we plan to update/maintain the library according to changes in ink! and contract pallet(You can see that part of our roadmap is a contribution to ink!).

3) We implemented a new analog of Erc1155(But ink! has already implemented that too), AccessControl, and Ownable from OpenZeppelin.
All implementations can be imported/injected/customized(you can override some logic like in Solidity) into another contract. While it is not possible with examples from ink!.
Each contract implementation has differences.
- Erc20 provide additional functionality(`mint`, `burn`, support of `Erc20Receiver`), `transfer` and `transfer_from` have been modified according our [PSP](https://github.com/w3f/PSPs/blob/master/PSPs/drafts/psp-22.md). All selectors generated according to the rules in ink! instead of Solidity. Changed error handling. So it is a native implementation of the PSP22 standard.
- Erc721 added `safe_transfer_from `. Support of metadata, Erc721Receiver. Changed error handling. Reduced the size of the compiled contract. In the future, we will also create PSP for non-fungible tokens.
All implementations are tested by unit tests and integration tests(real deployment of contracts to the node).
Added examples of how to use the library and customize default implementation.
https://github.com/Supercolony-net/openbrush-contracts/tree/main/examples

Noc2 reviewed (changes requested):
Thanks for the quick reply. I believe you should change the names of the contracts. Usually the name is assigned based on the github PR. So psp20 would refer to the 20 pr here https://github.com/w3f/PSPs/pulls, which might be confusing. Apart from that, I have one last question: Would you be willing to reduce the price of this grant a little bit? It seems a little bit expensive to me for the current deliverables.

themarkian commented:
@Noc2 yeah, the PSP that you merged recently was updated to psp22. We will update it on openbrush side just to be consistent. 

>Would you be willing to reduce the price of this grant a little bit? It seems a little bit expensive to me for the current deliverables.

Have an alternative suggestion. How about we add 1 more deliverable to the scope?

Noc2 commented:
Sounds good to me.

themarkian commented:
@Noc2 updated! :)

themarkian commented:
@Noc2 @BenWhiteJam @alxs @semuelle 
gentlemens, thanks for the thorough review and comments. And thanks for helping to make Openbrush successful 🚀
